### PR TITLE
Add git branching and trunk-based development doc

### PR DESCRIPTION
**Pull request by johnflavin**
Open to opinions and discussion on this.

### CODE REVIEW

File: docs/internal/git-branching-and-trunk-based-development.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Git Branching and Trunk-Based Development
+[Trunk-Based Development][https://trunkbaseddevelopment.com/] is a git branching strategy where the whole team integrates work into a "trunk" branch—`main` in our case. Changes are merged there frequently from short-lived feature branches. There are no other long-lived branches like a `develop` branch.
```
Comment (kathrynalpert) [inline docs/internal/git-branching-and-trunk-based-development.md:2]: Nit: fix link

**Review by kathrynalpert — COMMENTED** [review]
I believe the team concluded that independent QA is required before closing a ticket/releasing code. We want to increase the effort toward quality from the developer before handoff, but we felt that some level of "outside the box" thinking/testing could only be achieved through a separate set of eyes.

How do we reconcile this requirement on releasable code with the small, frequent, fast PRs you are suggesting here? Should we relax the constraint that `main` be releasable at any time, or should we require this QA before merging the PR (unless code is behind a feature flag)?

**Comment by johnflavin** [comment]
> I believe the team concluded that independent QA is required before closing a ticket/releasing code. We want to increase the effort toward quality from the developer before handoff, but we felt that some level of "outside the box" thinking/testing could only be achieved through a separate set of eyes.
> 
> How do we reconcile this requirement on releasable code with the small, frequent, fast PRs you are suggesting here? Should we relax the constraint that `main` be releasable at any time, or should we require this QA before merging the PR (unless code is behind a feature flag)?

First, yes to the last bit. If a feature is going to take a long time to develop, it should be disabled/feature flagged away. It could be developed and merged incrementally, but would not need to receive a thorough QA until it was "done" and ready to enable in production. During the whole development process it should still be subject to automated testing that enables the flag and runs however much is complete.

The way I was picturing QA is that we have automated testing that runs on every PR, every commit to `main`, maybe also daily, and we have a higher-effort manual QA that happens when a feature is fully "done" or when we are preparing for a release. I was not picturing a manual QA happening on every PR, necessarily.

I think the confusion might be that I used the word "releasable" in this document but I didn't define it. What I was thinking "releasable" meant is "not broken, the application(s) start and the tests pass", but it seems that what "releasable" means to you is—quite understandably—"fully complete and QAed and actually ready for release". If I change the wording to clarify what I'm trying to get at—that `main` should always be non-broken—would that be better?

**Comment by kathrynalpert** [comment]
> If I change the wording to clarify what I'm trying to get at—that `main` should always be non-broken—would that be better?

That would be great. I think also adding some more detail on how we intend to use feature flags (and we'll have a different PR merge/QA standard for this "work behind a feature flag") vs when we remove the feature flag and want to consider the feature "done."

File: docs/internal/git-branching-and-trunk-based-development.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+## Merge Conflict Resolution
+If my feature branch has a merge conflict with the trunk, how do I resolve it? There are two general strategies:
+1. Merge the trunk into the feature branch. You can then resolve the merge conflict in a merge commit on the feature branch.
+2. Rebase the feature branch onto the trunk. The commits on the feature branch are applied one-by-one onto the trunk, and you to fix any merge conflicts in each individual commit where a conflict occurs. This rewrites history, producing a "clean" feature branch which seems as if no merge conflict ever happened.
+
+It is the opinion of the author that option 2 is preferable to option 1. Rewriting history on a shared branch is fraught with peril, true, but on a branch that only a single person works on it is fine. The resulting clean commit history is worth the additional risk.
```
Comment (andylassiter) [inline docs/internal/git-branching-and-trunk-based-development.md:37]: I also like the clean commit history from option 2.

**Comment by andylassiter** [comment]
> > If I change the wording to clarify what I'm trying to get at—that `main` should always be non-broken—would that be better?
> 
> That would be great. I think also adding some more detail on how we intend to use feature flags (and we'll have a different PR merge/QA standard for this "work behind a feature flag") vs when we remove the feature flag and want to consider the feature "done."

More details on features flags would be helpful.

**Review by andylassiter — APPROVED** [review]

**Pull request merged**